xonsh/history.py

```diff
@@ -392,13 +431,15 @@ def _main(hist, args):
     if (args[0] == 'show' and len(args) > 1 and args[-1].startswith('-') and
             args[-1][1].isdigit()):
         args.insert(-1, '--')  # ensure parsing stops before a negative int
-    parser = _create_parser()
-    ns = parser.parse_args(args)
+    ns = _HIST_PARSER.parse_args(args)
     if ns.action is None:  # apply default action
-        ns = parser.parse_args(['show'] + args)
+        ns = _HIST_PARSER.parse_args(['show'] + args)
     _MAIN_ACTIONS[ns.action](ns, hist)
 
+
 def main(args=None, stdin=None):
     """This is the history command entry point."""
     _ = stdin
-    _main(builtins.__xonsh_history__, args)
+    _main(builtins.__xonsh_history__, args)  # pylint: disable=no-member
+
+_HIST_PARSER = _create_parser()
```

Comment:
I see your point and concur; although, see the timings below for objective data. I've switched to memoizing the function to avoid the need for the global statement and symbol.

I instrumented the history code as follows:

```
_s = time.time()
_HIST_PARSER = _create_parser()
print('WTF _create_parser {:f}'.format(time.time() - _s))
```

I also captured the starting time of main.py:

```
"""The main xonsh script."""
import time
_s = time.time()
```

Then reported the delta before entering the command loop:

```
ignore_sigtstp()
print('WTF about to shell.cmdloop {:f}'.format(time.time() - _s))
shell.cmdloop()
```

I then invoked xonsh five times. The average time to call _create_parser()
was 0.008793 seconds. The average time to present a shell prompt was
2.057667 seconds. So creating the history parser took 0.4% of the startup
time using a minimal XONSHRC that did little more than set $PATH.

Rerunning with no XONSHRC file (via `XONSHRC='' xonsh`) provided average
times of five runs of 0.010973 seconds for _create_parser() and 0.312528
seconds -- 3.5% of the total startup time.